csharp/CWE-760/HardcodedSalt.ql

```diff
@@ -0,0 +1,72 @@
+/**
+ * @name Hardcoded Salt
+ * @description Hardcoded Salt
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 6.1
+ * @precision medium
+ * @id cs/hardcoded-salt
+ * @tags security
+ *       external/cwe/cwe-760
+ */
+
+import csharp
+private import semmle.code.csharp.frameworks.Moq
+private import semmle.code.csharp.dataflow.DataFlow::DataFlow::PathGraph
+// import semmle.code.csharp.frameworks.system.security.Cryptography
+private import github.hardcoded
+private import github.crypto
+
+module HardcodedSalt {
+  abstract class Source extends DataFlow::ExprNode { }
+
+  abstract class Sink extends DataFlow::ExprNode { }
+
+  abstract class Sanitizer extends DataFlow::ExprNode { }
+
+  abstract class SanitizerGuard extends DataFlow::BarrierGuard { }
+
+  /*
+   * Sources
+   */
+
+  class Hardcoded extends Source {
+    Hardcoded() { this instanceof HardcodedValues }
+  }
+
+  /*
+   * Sinks
+   */
+
+  class HashAlgSalts extends Sink {
+    HashAlgSalts() { exists(Crypto::HashingAlgorithms hash | this = hash.getSalt()) }
+  }
+
+  /*
+   * Config
+   */
+
+  class TaintTrackingConfiguration extends TaintTracking::Configuration {
+    TaintTrackingConfiguration() { this = "HardcodedSalt" }
+
+    override predicate isSource(DataFlow::Node source) { source instanceof HardcodedSalt::Source }
+
+    override predicate isSink(DataFlow::Node sink) {
+      sink instanceof HardcodedSalt::Sink and
+      not any(ReturnedByMockObject mock).getAMemberInitializationValue() = sink.asExpr() and
+      not any(ReturnedByMockObject mock).getAnArgument() = sink.asExpr()
+    }
+
+    override predicate isSanitizer(DataFlow::Node node) { node instanceof Sanitizer }
```

Comment:
Can we not remove this line and line 25? If this was located in a .qll file, I would understand that it offers clients the ability to add things, but in a .ql file I would just leave it out. If it is ever needed, we can add it then.

Reply:
Ah, scratch that, I guess this does something, indeed.

Reply:
Still, lines 25 and 27 look a bit strange in this setting, I would remove them and directly refer to `DataFlow::ExprNode` and `DataFlow::BarrierGuard`